File: server/src/controllers/auth.controller.ts
import { Context } from "hono";
import { setCookie, deleteCookie, getCookie } from "hono/cookie";
import { AuthService } from "../services/auth.service";
import { loginSchema, registrationSchema, usernameSchema } from "../utils/validation";
import { sendSuccess, sendError } from "../utils/response";
import { verifyAccessToken, verifyRefreshToken, generateTokens } from "../utils/jwt";
import { publishRealtimeEvent, REALTIME_EVENTS_CHANNEL } from "../utils/realtimeEvents";

type LoginCredentials = {
  login: string;
  password: string;
};

type RegistrationCredentials = LoginCredentials & {
  username: string;
  publicKey: string;
  encryptedPrivateKey: string;
  encryptedPrivateKeyIv: string;
  encryptedPrivateKeySalt: string;
};

export class AuthController {
  constructor(
    private readonly authService: AuthService = new AuthService(),
    private readonly realtimeChannel: string = REALTIME_EVENTS_CHANNEL,
  ) {}

  private getFirstValidationErrorMessage(result: { error?: { issues?: Array<{ message?: string }> } }, fallback: string) {
    return result.error?.issues?.[0]?.message || fallback;
  }

  private isSecureRequest(c: Context): boolean {
    const forwardedProto = c.req.header("x-forwarded-proto")?.split(",")[0]?.trim().toLowerCase();
    if (forwardedProto) {
      return forwardedProto === "https";
    }

    return new URL(c.req.url).protocol === "https:";
  }

  private getRefreshCookieOptions(c: Context) {
    return {
      maxAge: 30 * 24 * 60 * 60,
      httpOnly: true,
      secure: this.isSecureRequest(c),
      sameSite: "strict" as const,
      path: "/",
    };
  }

  private async readJsonBody<T>(c: Context): Promise<T | null> {
    const contentType = c.req.header("content-type") ?? "";
    if (!contentType.includes("application/json")) {
      return null;
    }

    try {
      return await c.req.json<T>();
    } catch {
      return null;
    }
  }

  async register(c: Context) {
    const body = await this.readJsonBody<Partial<RegistrationCredentials>>(c);
    if (!body) {
      return sendError(c, 400, "Missing credentials");
    }

    const validatedData = registrationSchema.safeParse(body);
    if (!validatedData.success) {
      return sendError(c, 400, this.getFirstValidationErrorMessage(validatedData, "Invalid input data"));
    }

    const result = await this.authService.register(
      validatedData.data.login,
      validatedData.data.password,
      validatedData.data.username,
      validatedData.data.publicKey,
      validatedData.data.encryptedPrivateKey,
      validatedData.data.encryptedPrivateKeyIv,
      validatedData.data.encryptedPrivateKeySalt,
    );

    if ("error" in result) {
      return sendError(c, 401, result.error ?? "Registration failed");
    }

    setCookie(c, "refreshToken", result.tokens.refreshToken, this.getRefreshCookieOptions(c));

    return sendSuccess(c, "User registered successfully", {
      accessToken: result.tokens.accessToken,
    });
  }

  async login(c: Context) {
    const body = await this.readJsonBody<Partial<LoginCredentials>>(c);
    if (!body) {
      return sendError(c, 400, "Missing credentials");
    }

    const validatedData = loginSchema.safeParse(body);
    if (!validatedData.success) {
      return sendError(c, 400, this.getFirstValidationErrorMessage(validatedData, "Invalid input data"));
    }

    const result = await this.authService.login(validatedData.data.login, validatedData.data.password);

    if ("error" in result) {
      return sendError(c, 401, result.error ?? "Login failed");
    }

    setCookie(c, "refreshToken", result.tokens.refreshToken, this.getRefreshCookieOptions(c));

    return sendSuccess(c, "Login successful", {
      accessToken: result.tokens.accessToken,
      encryptedPrivateKey: result.encryptedPrivateKey,
      encryptedPrivateKeyIv: result.encryptedPrivateKeyIv,
      encryptedPrivateKeySalt: result.encryptedPrivateKeySalt,
    });
  }

  async refresh(c: Context) {
    const accessToken = c.req.header("authorization")?.split(" ")[1];
    const refreshToken = getCookie(c, "refreshToken");

    if (!refreshToken) {
      return sendError(c, 403, "Missing refresh token");
    }

    const refreshPayload = verifyRefreshToken(refreshToken);
    if (!refreshPayload) {
      return sendError(c, 403, "Invalid refresh token");
    }

    const user = await this.authService.getValidSessionUser(refreshPayload.userId, refreshToken);
    if (!user) {
      return sendError(c, 403, "Token mismatch");
    }

    const accessPayload = accessToken ? verifyAccessToken(accessToken) : null;
    if (accessPayload && accessPayload.userId !== user.id) {
      return sendError(c, 403, "Token mismatch");
    }

    const newTokens = generateTokens(user.id, user.username, user.role, user.login);
    await this.authService.saveRefreshToken(user.id, newTokens.refreshToken);

    setCookie(c, "refreshToken", newTokens.refreshToken, this.getRefreshCookieOptions(c));

    return sendSuccess(c, "Tokens refreshed", {
      accessToken: newTokens.accessToken,
    });
  }

  async logout(c: Context) {
    const accessToken = c.req.header("authorization")?.split(" ")[1];
    const refreshToken = getCookie(c, "refreshToken");
    let logoutUserId: number | null = null;

    const accessPayload = accessToken ? verifyAccessToken(accessToken) : null;
    const refreshPayload = refreshToken ? verifyRefreshToken(refreshToken) : null;

    if (accessPayload?.userId) {
      await this.authService.clearRefreshToken(accessPayload.userId);
      logoutUserId = accessPayload.userId;
    } else if (refreshPayload?.userId && refreshToken) {
      const user = await this.authService.getValidSessionUser(refreshPayload.userId, refreshToken);
      if (user) {
        await this.authService.clearRefreshToken(refreshPayload.userId);
        logoutUserId = refreshPayload.userId;
      }
    }

    if (logoutUserId) {
      await publishRealtimeEvent({
        type: "force_logout",
        userId: logoutUserId,
      }, undefined, this.realtimeChannel);
    }

    deleteCookie(c, "refreshToken", {
      path: "/",
    });

    return sendSuccess(c, "Logout successful");
  }

  async session(c: Context) {
    const accessToken = c.req.header("authorization")?.split(" ")[1];
    const refreshToken = getCookie(c, "refreshToken");

    if (!refreshToken) {
      return sendError(c, 403, "Session expired");
    }

    const refreshPayload = verifyRefreshToken(refreshToken);
    if (!refreshPayload) {
      return sendError(c, 403, "Session expired");
    }

    const user = await this.authService.getValidSessionUser(refreshPayload.userId, refreshToken);
    if (!user) {
      return sendError(c, 403, "Session expired");
    }

    const accessPayload = accessToken ? verifyAccessToken(accessToken) : null;
    if (accessPayload) {
      if (accessPayload.userId !== user.id) {
        return sendError(c, 403, "Session expired");
      }

      return sendSuccess(c, "Session valid", {
        accessToken,
      });
    }

    const newTokens = generateTokens(user.id, user.username, user.role, user.login);
    await this.authService.saveRefreshToken(user.id, newTokens.refreshToken);

    setCookie(c, "refreshToken", newTokens.refreshToken, this.getRefreshCookieOptions(c));

    return sendSuccess(c, "Session restored with new tokens", {
      accessToken: newTokens.accessToken,
    });
  }

  async updateUsername(c: Context) {
    const user = c.get("user");
    const body = await this.readJsonBody<Partial<{ username: string }>>(c);
    if (!body) {
      return sendError(c, 400, "Invalid username");
    }

    const validatedData = usernameSchema.safeParse(body);
    if (!validatedData.success) {
      return sendError(c, 400, this.getFirstValidationErrorMessage(validatedData, "Invalid username"));
    }

    const currentUser = await this.authService.getUserById(user.userId);
    if (!currentUser) {
      return sendError(c, 404, "User not found");
    }

    const normalizedUsername = validatedData.data.username.trim();
    if (currentUser.username === normalizedUsername) {
      return sendError(c, 409, "Username unchanged");
    }

    const updateResult = await this.authService.updateUsername(user.userId, normalizedUsername);

    if ("error" in updateResult) {
      return sendError(c, 409, updateResult.error);
    }

    const updatedUser = updateResult.user;
    if (!updatedUser) {
      return sendError(c, 500, "Failed to update username");
    }

    const newTokens = generateTokens(updatedUser.id, updatedUser.username, updatedUser.role, updatedUser.login);
    await this.authService.saveRefreshToken(updatedUser.id, newTokens.refreshToken);

    setCookie(c, "refreshToken", newTokens.refreshToken, this.getRefreshCookieOptions(c));

    return sendSuccess(c, "Username updated successfully", {
      accessToken: newTokens.accessToken,
      username: updatedUser.username,
    });
  }

  async getPublicKeys(c: Context) {
    const publicKeys = await this.authService.getAllPublicKeys();
    return sendSuccess(c, "Public keys retrieved", { publicKeys });
  }
}

File: server-go/controllers/auth_controller.go
package controllers

import (
	"context"
	"strconv"
	"strings"
	"time"

	"yourmsgr/config"
	"yourmsgr/realtime"
	"yourmsgr/services"
	"yourmsgr/utils"

	"github.com/gofiber/fiber/v2"
)

type AuthController struct {
	authService *services.AuthService
}

func NewAuthController() *AuthController {
	return &AuthController{
		authService: services.NewAuthService(),
	}
}

// Helper to determine if the request is secure (HTTPS)
func (ctrl *AuthController) isSecureRequest(c *fiber.Ctx) bool {
	forwardedProto := c.Get("X-Forwarded-Proto")
	if forwardedProto != "" {
		protos := strings.Split(forwardedProto, ",")
		return strings.TrimSpace(strings.ToLower(protos[0])) == "https"
	}
	return c.Protocol() == "https"
}

// Helper to get refresh token cookie options
func (ctrl *AuthController) getRefreshCookieOptions(c *fiber.Ctx) *fiber.Cookie {
	return &fiber.Cookie{
		Name:     "refreshToken",
		HTTPOnly: true,
		Secure:   ctrl.isSecureRequest(c),
		SameSite: "Strict",
		Path:     "/",
		Expires:  time.Now().Add(30 * 24 * time.Hour), // 30 days
	}
}

// JWTAuthMiddleware verifies the JWT access token and sets user payload in Context locals
func JWTAuthMiddleware() fiber.Handler {
	return func(c *fiber.Ctx) error {
		authHeader := c.Get("Authorization")
		if authHeader == "" {
			return utils.SendError(c, fiber.StatusUnauthorized, "Unauthorized")
		}

		parts := strings.Split(authHeader, " ")
		if len(parts) != 2 || strings.ToLower(parts[0]) != "bearer" {
			return utils.SendError(c, fiber.StatusUnauthorized, "Unauthorized")
		}

		tokenString := parts[1]
		claims, err := utils.VerifyToken(tokenString, config.AppConfig.JWTAccessSecret)
		if err != nil {
			return utils.SendError(c, fiber.StatusUnauthorized, "Invalid or expired token")
		}

		// Store user info in Fiber context locals
		c.Locals("userId", claims.UserID)
		c.Locals("username", claims.UserName)
		c.Locals("userRole", claims.UserRole)
		c.Locals("login", claims.Login)

		return c.Next()
	}
}

// RegisterRequest represents the payload for user registration
type RegisterRequest struct {
	Login                   string `json:"login"`
	Password                string `json:"password"`
	Username                string `json:"username"`
	PublicKey               string `json:"publicKey"`
	EncryptedPrivateKey     string `json:"encryptedPrivateKey"`
	EncryptedPrivateKeyIv   string `json:"encryptedPrivateKeyIv"`
	EncryptedPrivateKeySalt string `json:"encryptedPrivateKeySalt"`
}

func (ctrl *AuthController) Register(c *fiber.Ctx) error {
	var req RegisterRequest
	if err := c.BodyParser(&req); err != nil {
		return utils.SendError(c, fiber.StatusBadRequest, "Invalid body")
	}

	// Validate inputs
	req.Login = strings.TrimSpace(req.Login)
	req.Username = strings.TrimSpace(req.Username)
	if req.Login == "" || req.Password == "" || req.Username == "" {
		return utils.SendError(c, fiber.StatusBadRequest, "Missing credentials")
	}

	if len(req.Login) < 3 || len(req.Login) > 20 {
		return utils.SendError(c, fiber.StatusBadRequest, "Login must be between 3 and 20 characters")
	}
	if len(req.Username) < 3 || len(req.Username) > 20 {
		return utils.SendError(c, fiber.StatusBadRequest, "Username must be between 3 and 20 characters")
	}
	if len(req.Password) < 6 {
		return utils.SendError(c, fiber.StatusBadRequest, "Password must be at least 6 characters")
	}

	ctx := context.Background()
	user, err := ctrl.authService.Register(ctx, req.Login, req.Password, req.Username,
		req.PublicKey, req.EncryptedPrivateKey, req.EncryptedPrivateKeyIv, req.EncryptedPrivateKeySalt)

	if err != nil {
		return utils.SendError(c, fiber.StatusUnauthorized, err.Error())
	}

	// Generate tokens
	accessToken, refreshToken, err := utils.GenerateTokens(user.ID, user.Username, user.Role, user.Login,
		config.AppConfig.JWTAccessSecret, config.AppConfig.JWTRefreshSecret)
	if err != nil {
		return utils.SendError(c, fiber.StatusInternalServerError, "Failed to generate tokens")
	}

	// Save hashed refresh token to DB
	hashedRefresh := utils.HashRefreshToken(refreshToken, config.AppConfig.JWTRefreshSecret)
	if err := ctrl.authService.SaveRefreshToken(ctx, user.ID, hashedRefresh); err != nil {
		return utils.SendError(c, fiber.StatusInternalServerError, "Failed to save session")
	}

	// Set refresh token in cookie
	cookie := ctrl.getRefreshCookieOptions(c)
	cookie.Value = refreshToken
	c.Cookie(cookie)

	return utils.SendSuccess(c, "User registered successfully", fiber.Map{
		"accessToken": accessToken,
	})
}

// LoginRequest represents the payload for user login
type LoginRequest struct {
	Login    string `json:"login"`
	Password string `json:"password"`
}

func (ctrl *AuthController) Login(c *fiber.Ctx) error {
	var req LoginRequest
	if err := c.BodyParser(&req); err != nil {
		return utils.SendError(c, fiber.StatusBadRequest, "Invalid body")
	}

	req.Login = strings.TrimSpace(req.Login)
	if req.Login == "" || req.Password == "" {
		return utils.SendError(c, fiber.StatusBadRequest, "Missing credentials")
	}

	ctx := context.Background()
	user, err := ctrl.authService.Login(ctx, req.Login, req.Password)
	if err != nil {
		return utils.SendError(c, fiber.StatusUnauthorized, err.Error())
	}

	// Generate tokens
	accessToken, refreshToken, err := utils.GenerateTokens(user.ID, user.Username, user.Role, user.Login,
		config.AppConfig.JWTAccessSecret, config.AppConfig.JWTRefreshSecret)
	if err != nil {
		return utils.SendError(c, fiber.StatusInternalServerError, "Failed to generate tokens")
	}

	// Save hashed refresh token to DB
	hashedRefresh := utils.HashRefreshToken(refreshToken, config.AppConfig.JWTRefreshSecret)
	if err := ctrl.authService.SaveRefreshToken(ctx, user.ID, hashedRefresh); err != nil {
		return utils.SendError(c, fiber.StatusInternalServerError, "Failed to save session")
	}

	// Set refresh token in cookie
	cookie := ctrl.getRefreshCookieOptions(c)
	cookie.Value = refreshToken
	c.Cookie(cookie)

	return utils.SendSuccess(c, "Login successful", fiber.Map{
		"accessToken":             accessToken,
		"encryptedPrivateKey":     user.EncryptedPrivateKey,
		"encryptedPrivateKeyIv":   user.EncryptedPrivateKeyIv,
		"encryptedPrivateKeySalt": user.EncryptedPrivateKeySalt,
	})
}

func (ctrl *AuthController) Refresh(c *fiber.Ctx) error {
	authHeader := c.Get("Authorization")
	refreshToken := c.Cookies("refreshToken")

	if refreshToken == "" {
		return utils.SendError(c, fiber.StatusForbidden, "Missing refresh token")
	}

	// Verify refresh token
	refreshPayload, err := utils.VerifyToken(refreshToken, config.AppConfig.JWTRefreshSecret)
	if err != nil {
		return utils.SendError(c, fiber.StatusForbidden, "Invalid refresh token")
	}

	ctx := context.Background()
	user, err := ctrl.authService.GetValidSessionUser(ctx, refreshPayload.UserID, refreshToken, config.AppConfig.JWTRefreshSecret)
	if err != nil || user == nil {
		return utils.SendError(c, fiber.StatusForbidden, "Token mismatch")
	}

	// Match access token ownership if present
	if authHeader != "" {
		parts := strings.Split(authHeader, " ")
		if len(parts) == 2 && strings.ToLower(parts[0]) == "bearer" {
			accessPayload, err := utils.VerifyToken(parts[1], config.AppConfig.JWTAccessSecret)
			if err == nil && accessPayload.UserID != user.ID {
				return utils.SendError(c, fiber.StatusForbidden, "Token mismatch")
			}
		}
	}

	// Generate new tokens
	newAccessToken, newRefreshToken, err := utils.GenerateTokens(user.ID, user.Username, user.Role, user.Login,
		config.AppConfig.JWTAccessSecret, config.AppConfig.JWTRefreshSecret)
	if err != nil {
		return utils.SendError(c, fiber.StatusInternalServerError, "Failed to generate tokens")
	}

	// Save new hashed refresh token (rotate old session token)
	hashedRefresh := utils.HashRefreshToken(newRefreshToken, config.AppConfig.JWTRefreshSecret)
	if err := ctrl.authService.RotateRefreshToken(ctx, user.ID, refreshToken, hashedRefresh, config.AppConfig.JWTRefreshSecret); err != nil {
		return utils.SendError(c, fiber.StatusInternalServerError, "Failed to save session")
	}

	// Set new cookie
	cookie := ctrl.getRefreshCookieOptions(c)
	cookie.Value = newRefreshToken
	c.Cookie(cookie)

	return utils.SendSuccess(c, "Tokens refreshed", fiber.Map{
		"accessToken": newAccessToken,
	})
}

func (ctrl *AuthController) Logout(c *fiber.Ctx) error {
	authHeader := c.Get("Authorization")
	refreshToken := c.Cookies("refreshToken")
	var logoutUserId int64

	ctx := context.Background()

	// 1. Try to fetch user from Access Token
	if authHeader != "" {
		parts := strings.Split(authHeader, " ")
		if len(parts) == 2 && strings.ToLower(parts[0]) == "bearer" {
			accessPayload, err := utils.VerifyToken(parts[1], config.AppConfig.JWTAccessSecret)
			if err == nil {
				logoutUserId = accessPayload.UserID
			}
		}
	}

	// 2. If access token is invalid, try Refresh Token
	if logoutUserId == 0 && refreshToken != "" {
		refreshPayload, err := utils.VerifyToken(refreshToken, config.AppConfig.JWTRefreshSecret)
		if err == nil {
			user, err := ctrl.authService.GetValidSessionUser(ctx, refreshPayload.UserID, refreshToken, config.AppConfig.JWTRefreshSecret)
			if err == nil && user != nil {
				logoutUserId = user.ID
			}
		}
	}

	// 3. Clear token in DB
	if logoutUserId > 0 {
		if refreshToken != "" {
			ctrl.authService.RemoveRefreshToken(ctx, logoutUserId, refreshToken, config.AppConfig.JWTRefreshSecret)
		} else {
			ctrl.authService.ClearAllRefreshTokens(ctx, logoutUserId)
			realtime.PublishRealtimeEvent(ctx, realtime.RealtimeEvent{
				Type:   "force_logout",
				UserID: logoutUserId,
			})
		}
	}

	// 4. Delete refresh token cookie
	c.ClearCookie("refreshToken")

	return utils.SendSuccess(c, "Logout successful", nil)
}

func (ctrl *AuthController) Session(c *fiber.Ctx) error {
	authHeader := c.Get("Authorization")
	refreshToken := c.Cookies("refreshToken")

	if refreshToken == "" {
		return utils.SendError(c, fiber.StatusForbidden, "Session expired")
	}

	refreshPayload, err := utils.VerifyToken(refreshToken, config.AppConfig.JWTRefreshSecret)
	if err != nil {
		return utils.SendError(c, fiber.StatusForbidden, "Session expired")
	}

	ctx := context.Background()
	user, err := ctrl.authService.GetValidSessionUser(ctx, refreshPayload.UserID, refreshToken, config.AppConfig.JWTRefreshSecret)
	if err != nil || user == nil {
		return utils.SendError(c, fiber.StatusForbidden, "Session expired")
	}

	if authHeader != "" {
		parts := strings.Split(authHeader, " ")
		if len(parts) == 2 && strings.ToLower(parts[0]) == "bearer" {
			accessPayload, err := utils.VerifyToken(parts[1], config.AppConfig.JWTAccessSecret)
			if err == nil {
				if accessPayload.UserID != user.ID {
					return utils.SendError(c, fiber.StatusForbidden, "Session expired")
				}
				return utils.SendSuccess(c, "Session valid", fiber.Map{
					"accessToken": parts[1],
				})
			}
		}
	}

	// Restore session with new tokens
	newAccessToken, newRefreshToken, err := utils.GenerateTokens(user.ID, user.Username, user.Role, user.Login,
		config.AppConfig.JWTAccessSecret, config.AppConfig.JWTRefreshSecret)
	if err != nil {
		return utils.SendError(c, fiber.StatusInternalServerError, "Failed to restore session")
	}

	hashedRefresh := utils.HashRefreshToken(newRefreshToken, config.AppConfig.JWTRefreshSecret)
	if err := ctrl.authService.SaveRefreshToken(ctx, user.ID, hashedRefresh); err != nil {
		return utils.SendError(c, fiber.StatusInternalServerError, "Failed to save session")
	}

	cookie := ctrl.getRefreshCookieOptions(c)
	cookie.Value = newRefreshToken
	c.Cookie(cookie)

	return utils.SendSuccess(c, "Session restored with new tokens", fiber.Map{
		"accessToken": newAccessToken,
	})
}

type UpdateUsernameRequest struct {
	Username string `json:"username"`
}

func (ctrl *AuthController) UpdateUsername(c *fiber.Ctx) error {
	userId := c.Locals("userId").(int64)

	var req UpdateUsernameRequest
	if err := c.BodyParser(&req); err != nil {
		return utils.SendError(c, fiber.StatusBadRequest, "Invalid body")
	}

	req.Username = strings.TrimSpace(req.Username)
	if req.Username == "" || len(req.Username) < 3 || len(req.Username) > 20 {
		return utils.SendError(c, fiber.StatusBadRequest, "Invalid username")
	}

	ctx := context.Background()
	user, err := ctrl.authService.GetUserById(ctx, userId)
	if err != nil || user == nil {
		return utils.SendError(c, fiber.StatusNotFound, "User not found")
	}

	if user.Username == req.Username {
		return utils.SendError(c, fiber.StatusConflict, "Username unchanged")
	}

	updatedUser, affectedUserIds, err := ctrl.authService.UpdateUsername(ctx, userId, req.Username)
	if err != nil {
		return utils.SendError(c, fiber.StatusConflict, err.Error())
	}

	// Generate new tokens
	accessToken, refreshToken, err := utils.GenerateTokens(updatedUser.ID, updatedUser.Username, updatedUser.Role, updatedUser.Login,
		config.AppConfig.JWTAccessSecret, config.AppConfig.JWTRefreshSecret)
	if err != nil {
		return utils.SendError(c, fiber.StatusInternalServerError, "Failed to generate tokens")
	}

	// Invalidate all active sessions because username has changed (embedded in JWT payload)
	ctrl.authService.ClearAllRefreshTokens(ctx, updatedUser.ID)

	hashedRefresh := utils.HashRefreshToken(refreshToken, config.AppConfig.JWTRefreshSecret)
	if err := ctrl.authService.SaveRefreshToken(ctx, updatedUser.ID, hashedRefresh); err != nil {
		return utils.SendError(c, fiber.StatusInternalServerError, "Failed to save session")
	}

	cookie := ctrl.getRefreshCookieOptions(c)
	cookie.Value = refreshToken
	c.Cookie(cookie)

	// Publish realtime sync and force_logout events for scaling
	realtime.PublishRealtimeEvent(ctx, realtime.RealtimeEvent{
		Type: "sync_group_messages",
	})
	realtime.PublishRealtimeEvent(ctx, realtime.RealtimeEvent{
		Type:   "force_logout",
		UserID: updatedUser.ID,
	})
	if len(affectedUserIds) > 0 {
		realtime.PublishRealtimeEvent(ctx, realtime.RealtimeEvent{
			Type:    "sync_private_chats",
			UserIDs: affectedUserIds,
		})
	}

	return utils.SendSuccess(c, "Username updated successfully", fiber.Map{
		"accessToken": accessToken,
		"username":    updatedUser.Username,
	})
}

func (ctrl *AuthController) GetPublicKeys(c *fiber.Ctx) error {
	userId := c.Locals("userId").(int64)
	rawUserIds := c.Query("userIds")
	var targetUserIds []int64

	if rawUserIds != "" {
		parts := strings.Split(rawUserIds, ",")
		for _, part := range parts {
			val, err := strconv.ParseInt(strings.TrimSpace(part), 10, 64)
			if err == nil && val > 0 {
				targetUserIds = append(targetUserIds, val)
			}
		}

		if len(targetUserIds) == 0 {
			return utils.SendError(c, fiber.StatusBadRequest, "Invalid userIds query")
		}

		if len(targetUserIds) > 50 {
			return utils.SendError(c, fiber.StatusBadRequest, "Too many requested userIds")
		}
	}

	ctx := context.Background()
	publicKeys, err := ctrl.authService.GetPublicKeysForUser(ctx, userId, targetUserIds)
	if err != nil {
		return utils.SendError(c, fiber.StatusInternalServerError, "Failed to retrieve public keys")
	}

	return utils.SendSuccess(c, "Public keys retrieved", fiber.Map{
		"publicKeys": publicKeys,
	})
}
